shared/src/ledger/eth_bridge/storage/bridge_pool.rs

```diff
@@ -44,3 +58,847 @@ pub fn is_bridge_pool_key(key: &Key) -> bool {
 pub fn is_protected_storage(key: &Key) -> bool {
     is_bridge_pool_key(key) && *key != get_pending_key()
 }
+
+/// A simple Merkle tree for the Ethereum bridge pool
+#[derive(
+    Debug, Default, Clone, BorshSerialize, BorshDeserialize, BorshSchema,
+)]
+pub struct BridgePoolTree {
+    /// Root of the tree
+    root: KeccakHash,
+    /// The underlying storage
+    store: BTreeSet<KeccakHash>,
+}
+
+impl BridgePoolTree {
+    /// Create a new merkle tree for the Ethereum bridge pool
+    pub fn new(root: KeccakHash, store: BTreeSet<KeccakHash>) -> Self {
+        Self { root, store }
+    }
+
+    /// Parse the key to ensure it is of the correct type.
+    ///
+    /// If it is, it can be converted to a hash.
+    /// Checks if the hash is in the tree.
+    pub fn contains_key(&self, key: &Key) -> Result<bool, Error> {
+        Ok(self.store.contains(&Self::parse_key(key)?))
+    }
+
+    /// Update the tree with a new value.
+    ///
+    /// Returns the new root if successful. Will
+    /// return an error if the key is malformed.
+    pub fn update_key(&mut self, key: &Key) -> Result<Hash, Error> {
```

Comment:
nit: would call this `insert_key` rather than `update_key`

Reply:
We can overwrite an existing key in theory, but that would imply a hash collision so that would never happen in practice. I can change it though.

Reply:
Yeah if it should never happen, logging here at `tracing::error` level would also document nicely that that should never be the case (and if it does happen somehow, we can see what happened)